Importing libraries

In [47]:
import requests
from bs4 import BeautifulSoup
import csv
import json
import lxml
import pandas as pd

Scrape data from Wikipedia

In [28]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text


The tabular data is availabe in table and belongs to class="wikitable sortable"So let's extract only table

In [32]:
from bs4 import BeautifulSoup

soup = BeautifulSoup(website_url,'html.parser')

In [34]:
neighborhood = soup.find('table', class_ = 'wikitable')
neighborhood_rows = neighborhood.find_all('tr')

Extract the columns 'Postcode', 'Borough', 'Neighbourhood' from the table

In [35]:
information = []
for row in neighborhood_rows:
    info = row.text.split('\n')[1:-1] # remove empty str (the first and last items)
    information.append(info)
    
information[0:20] #preview the first 20 rows

[['Postal Code', '', 'Borough', '', 'Neighborhood'],
 ['M1A', '', 'Not assigned', '', 'Not assigned'],
 ['M2A', '', 'Not assigned', '', 'Not assigned'],
 ['M3A', '', 'North York', '', 'Parkwoods'],
 ['M4A', '', 'North York', '', 'Victoria Village'],
 ['M5A', '', 'Downtown Toronto', '', 'Regent Park, Harbourfront'],
 ['M6A', '', 'North York', '', 'Lawrence Manor, Lawrence Heights'],
 ['M7A',
  '',
  'Downtown Toronto',
  '',
  "Queen's Park, Ontario Provincial Government"],
 ['M8A', '', 'Not assigned', '', 'Not assigned'],
 ['M9A', '', 'Etobicoke', '', 'Islington Avenue, Humber Valley Village'],
 ['M1B', '', 'Scarborough', '', 'Malvern, Rouge'],
 ['M2B', '', 'Not assigned', '', 'Not assigned'],
 ['M3B', '', 'North York', '', 'Don Mills'],
 ['M4B', '', 'East York', '', 'Parkview Hill, Woodbine Gardens'],
 ['M5B', '', 'Downtown Toronto', '', 'Garden District, Ryerson'],
 ['M6B', '', 'North York', '', 'Glencairn'],
 ['M7B', '', 'Not assigned', '', 'Not assigned'],
 ['M8B', '', 'Not assigne

Turn above information table into a Pandas dataframe


In [38]:
neighbor_df = pd.DataFrame(information[1:], columns=information[0])
# where information[1:] contains each row of neighborhoods
# and columns = information[0] gives the column names


# Re-spell Neighbourhood as Neighborhood
#neighbor_df = neighbor_df.rename(columns={neighbor_df.columns[2]: "Neighborhood" })


neighbor_df.head(20)

Postal Code             Borough    \
0          M1A        Not assigned     
1          M2A        Not assigned     
2          M3A          North York     
3          M4A          North York     
4          M5A    Downtown Toronto     
5          M6A          North York     
6          M7A    Downtown Toronto     
7          M8A        Not assigned     
8          M9A           Etobicoke     
9          M1B         Scarborough     
10         M2B        Not assigned     
11         M3B          North York     
12         M4B           East York     
13         M5B    Downtown Toronto     
14         M6B          North York     
15         M7B        Not assigned     
16         M8B        Not assigned     
17         M9B           Etobicoke     
18         M1C         Scarborough     
19         M2C        Not assigned     

                                         Neighborhood  
0                                        Not assigned  
1                                        Not assigned  
2                                           Parkwoods  
3                                    Victoria Village  
4                           Regent Park, Harbourfront  
5                    Lawrence Manor, Lawrence Heights  
6         Queen's Park, Ontario Provincial Government  
7                                        Not assigned  
8             Islington Avenue, Humber Valley Village  
9                                      Malvern, Rouge  
10                                       Not assigned  
11                                          Don Mills  
12                    Parkview Hill, Woodbine Gardens  
13                           Garden District, Ryerson  
14                                          Glencairn  
15                                       Not assigned  
16                                       Not assigned  
17  West Deane Park, Princess Gardens, Martin Grov...  
18             Rouge Hill, Port Union, Highland Creek  
19                                       Not assigned

Cleaning acc to conditions

In [39]:
neighbor_df = neighbor_df[neighbor_df.Borough != 'Not assigned']

neighbor_df.reset_index(drop=True, inplace=True)

neighbor_df.head(20)


Postal Code             Borough    \
0          M3A          North York     
1          M4A          North York     
2          M5A    Downtown Toronto     
3          M6A          North York     
4          M7A    Downtown Toronto     
5          M9A           Etobicoke     
6          M1B         Scarborough     
7          M3B          North York     
8          M4B           East York     
9          M5B    Downtown Toronto     
10         M6B          North York     
11         M9B           Etobicoke     
12         M1C         Scarborough     
13         M3C          North York     
14         M4C           East York     
15         M5C    Downtown Toronto     
16         M6C                York     
17         M9C           Etobicoke     
18         M1E         Scarborough     
19         M4E        East Toronto     

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...  
12             Rouge Hill, Port Union, Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights  
15                                     St. James Town  
16                                 Humewood-Cedarvale  
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...  
18                  Guildwood, Morningside, West Hill  
19                                        The Beaches

Combine neighborhoods with the same postcode


In [46]:
grouped = neighbor_df.groupby(['Postal Code']) # group by Postcode


# combine the neighborhoods grouped by postcode and into a new df
neighborhood_grouped = grouped['Neighborhood'].apply(lambda x: x.sum()) 
# adds spaces and commas between neighborhoods
neighborhood_grouped = grouped['Neighborhood'].apply(lambda x: "%s" % ', '.join(x))
# matches a borough to each postcode
borough_grouped = grouped['Borough'].apply(lambda x: set(x).pop())
# turn borough_grouped and neighborhood_grouped into dataframes
borough = borough_grouped.to_frame()
neighborhood = neighborhood_grouped.to_frame()
#combine the dataframe borough and the dataframe neighborhood into one dataframe
grouped_final = borough.merge(neighborhood, on="Postal Code")

grouped_final

Borough                                       Neighborhood
Postal Code                                                                
M1B          Scarborough                                     Malvern, Rouge
M1C          Scarborough             Rouge Hill, Port Union, Highland Creek
M1E          Scarborough                  Guildwood, Morningside, West Hill
M1G          Scarborough                                             Woburn
M1H          Scarborough                                          Cedarbrae
...                  ...                                                ...
M9N                 York                                             Weston
M9P            Etobicoke                                          Westmount
M9R            Etobicoke  Kingsview Village, St. Phillips, Martin Grove ...
M9V            Etobicoke  South Steeles, Silverstone, Humbergate, Jamest...
M9W            Etobicoke                Northwest, West Humber - Clairville

[103 rows x 2 columns]

Print the .shape method to print the number of rows of the dataframe


In [50]:
print('The number of rows and columns in this final grouped dataframe is',grouped_final.shape)

The number of rows and columns in this final grouped dataframe is (103, 2)


Attaching longitudes and latitudes

In [54]:
geospatial_data = pd.read_csv(r'C:\Users\aksha\OneDrive\Desktop\ibm\Geospatial_Coordinates.csv')

geospatial_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [60]:
geospatial_data = geospatial_data.rename(columns={geospatial_data.columns[0]: "Postal Code" })

In [61]:
geospatial_data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [63]:
full_table = grouped_final.merge(geospatial_data, on = 'Postal Code')

full_table.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [64]:
full_table

Postal Code      Borough  \
0           M1B  Scarborough   
1           M1C  Scarborough   
2           M1E  Scarborough   
3           M1G  Scarborough   
4           M1H  Scarborough   
..          ...          ...   
98          M9N         York   
99          M9P    Etobicoke   
100         M9R    Etobicoke   
101         M9V    Etobicoke   
102         M9W    Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724  
101  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
102                Northwest, West Humber - Clairville  43.706748 -79.594054  

[103 rows x 5 columns]

Explore and cluster the neighborhoods of Toronto

Importing all the necessary libraries

In [65]:
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 
import requests 
from pandas.io.json import json_normalize 

 TO see how many boroughs and neighborhoods are in Toronto


In [66]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(full_table['Borough'].unique()),
        full_table['Neighborhood'].shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


get the longitude / latitude of toronto

In [67]:
address = 'Toronto, Ontario, Canada'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Create a map of Toronto with neighborhoods superimposed on top

In [69]:
import folium
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 10)
for lat, long, bor, neigh in zip(full_table['Latitude'], full_table['Longitude'], 
                                 full_table['Borough'], full_table['Neighborhood']):
    label = '{}, {}'.format(neigh, bor)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius = 7, 
        popup = label,
        color = 'red',
        fill = True,
        fill_color = 'white',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_toronto)
        
map_toronto


The map above shows all 103 neighborhoods in Toronto, using Folium map and a circle marker in the colors of the Canadian flag. Clicking over the markers, a label shows the neighborhood and the borough names.